File: chapter06/task-injection/python/report_server.py
# A stand-in for any long-running backend: start_report returns immediately with a
# handle, the actual work finishes a few seconds later. With the Tasks extension the
# server would return a CreateTaskResult instead of a handle-in-content, but the
# host-side injection problem this sample demonstrates is identical either way.
import asyncio
import json
import os

from mcp.server import MCPServer

server = MCPServer(name="report-server", version="0.1.0")

# None = still running; str = the finished report.
_results: dict[str, str | None] = {}
_counter = 0
_pending: set[asyncio.Task[None]] = set()

DELAY_SECONDS = float(os.environ.get("REPORT_DELAY_SECONDS", "3"))


@server.tool()
async def start_report(topic: str) -> str:
    """Start generating a report. Returns immediately with an operation handle; the report completes a few seconds later.

    Args:
        topic: What the report should cover.
    """
    global _counter
    _counter += 1
    operation_id = f"op-{_counter:03d}"
    _results[operation_id] = None

    async def finish() -> None:
        await asyncio.sleep(DELAY_SECONDS)
        _results[operation_id] = (
            f"Report on {topic}: revenue up 14% quarter over quarter, "
            "driven by the new subscription tier. Two regions missed target; details in the appendix."
        )

    task = asyncio.create_task(finish())
    _pending.add(task)
    task.add_done_callback(_pending.discard)
    return json.dumps({"operationId": operation_id, "status": "running"})


# The parameter is camelCase so the tool contract on the wire is identical to the
# C# and TypeScript ports.
@server.tool()
def get_report_result(operationId: str) -> str:
    """Fetch the result of a previously started report. Returns status 'running' until it completes.

    Args:
        operationId: Handle returned by start_report.
    """
    if operationId not in _results:
        return json.dumps({"operationId": operationId, "status": "unknown"})
    report = _results[operationId]
    if report is None:
        return json.dumps({"operationId": operationId, "status": "running"})
    return json.dumps({"operationId": operationId, "status": "completed", "report": report})


if __name__ == "__main__":
    server.run()

File: chapter06/task-injection/typescript/src/report-server.ts
// A stand-in for any long-running backend: start_report returns immediately with a
// handle, the actual work finishes a few seconds later. With the Tasks extension the
// server would return a CreateTaskResult instead of a handle-in-content, but the
// host-side injection problem this sample demonstrates is identical either way.
import { McpServer } from '@modelcontextprotocol/server';
import { serveStdio } from '@modelcontextprotocol/server/stdio';
import { z } from 'zod';

// null = still running; string = the finished report. Module-level, so the state is
// shared even if serveStdio pins separate era instances to the connection.
const results = new Map<string, string | null>();
let counter = 0;

const delayMs = 1000 * Number(process.env.REPORT_DELAY_SECONDS ?? '3');

const jsonResult = (value: object) => ({
  content: [{ type: 'text' as const, text: JSON.stringify(value) }],
});

// serveStdio owns the era decision: modern (2026-07-28) openings get a modern
// instance, legacy openings a legacy one — the same factory serves both.
serveStdio(() => {
  const server = new McpServer({ name: 'report-server', version: '0.1.0' });

  server.registerTool(
    'start_report',
    {
      description:
        'Start generating a report. Returns immediately with an operation handle; ' +
        'the report completes a few seconds later.',
      inputSchema: z.object({
        topic: z.string().describe('What the report should cover.'),
      }),
    },
    async ({ topic }) => {
      const operationId = `op-${String(++counter).padStart(3, '0')}`;
      results.set(operationId, null);
      setTimeout(() => {
        results.set(
          operationId,
          `Report on ${topic}: revenue up 14% quarter over quarter, ` +
            'driven by the new subscription tier. Two regions missed target; details in the appendix.',
        );
      }, delayMs);
      return jsonResult({ operationId, status: 'running' });
    },
  );

  server.registerTool(
    'get_report_result',
    {
      description:
        "Fetch the result of a previously started report. Returns status 'running' until it completes.",
      inputSchema: z.object({
        operationId: z.string().describe('Handle returned by start_report.'),
      }),
    },
    async ({ operationId }) => {
      if (!results.has(operationId)) return jsonResult({ operationId, status: 'unknown' });
      const report = results.get(operationId);
      return report == null
        ? jsonResult({ operationId, status: 'running' })
        : jsonResult({ operationId, status: 'completed', report });
    },
  );

  return server;
});
